### PR TITLE
Implement Harvest abort button

### LINKED ISSUE #1991: Add abort button to various upload screens
**Is your feature request related to a problem? Please describe.**

The harvests API supports aborting a harvest. Yet that is not exposed in the client anywhere.

**Describe the solution you'd like**

Add a "Cancel" button that aborts the harvest (by patching the harvest's status as `complete`) on the 

- harvest list page
- harvest upload screen
- and metadata review screen

**Describe alternatives you've considered**

N/A

**Additional context**

### PR DESCRIPTION
**Pull request by hudson-newey**
# Implement Harvest Abort Button on Various Screens

What is the purpose of this PR?
This PR creates and implements an "Abort" button and corresponding modal on the harvest upload list page, batch upload screen, and metadata review screens.

## Changes

* Creates an "Abort" button and modal
* Creates transition() method in list.component.ts

## Problems

N/A

## Issues

Fixes #1991

## Visual Changes

Project harvest list view
image

Harvest upload screen
image

Harvest metadata review screen
image

Abort conformation modal
image

## Final Checklist

- [ ] Assign reviewers if you have permission
- [ ] Assign labels if you have permission
- [ ] Link issues related to PR
- [ ] Ensure project linter is not producing any warnings (`npm run lint`)
- [ ] Ensure build is passing on all browsers (`npm run test:all`)
- [ ] Ensure CI build is passing
- [ ] Ensure docker container is passing ([docs](https://github.com/QutEcoacoustics/workbench-client#docker))

### CODE REVIEW

File: src/app/components/harvest/pages/list/list.component.html [hunk 1]
```diff
@@ -42,6 +42,26 @@ <h1>Upload History</h1>
       <a class="btn btn-sm btn-primary" [bawUrl]="row.viewUrl">
         {{ asHarvest(row).status !== "complete" ? "Continue" : "View" }}
       </a>
+      <a
+        *ngIf="asHarvest(row).status !== 'complete'"
```
Comment (atruskie) [inline src/app/components/harvest/pages/list/list.component.html]: move this check to a function

it also needs to be a more complex check: as per https://github.com/QutEcoacoustics/baw-server/blob/0e54cbd30ae9fc704466cce11fbd1ba4de8e674e/app/models/harvest.rb#L395-L399 updating a harvest is only allowed in certain states.

File: src/app/components/harvest/pages/list/list.component.ts [hunk 2]
```diff
@@ -44,6 +44,55 @@ class ListComponent extends PageComponent implements OnInit {
   public ngOnInit(): void {
     this.project = this.route.snapshot.data[projectKey].model;
     this.canCreateHarvestCapability = this.project.can("createHarvest").can;
+    this.filters$ = new BehaviorSubject({
```
Comment (atruskie) [inline src/app/components/harvest/pages/list/list.component.ts:53]: add a comment explaining why a whole BehaviorSubject is needed for a seemingly static filter

File: src/app/components/harvest/pages/list/list.component.ts [hunk 3]
```diff
@@ -44,6 +44,55 @@ class ListComponent extends PageComponent implements OnInit {
...
+            sorting: {
+              direction: "desc",
+              orderBy: "createdAt"
+            }
+          });
+        },
+        error: (err: BawApiError): void => {
+          this.notifications.error(err.message);
+        },
+      });
```
Comment (atruskie) [inline src/app/components/harvest/pages/list/list.component.ts]: massively simplify this.

The only function you need is an `abortHarvest` function.  Copying and pasting this wholesale is a bad idea.

**Review by atruskie — CHANGES_REQUESTED** [review]
Missing tests for the metadata review screen and the list page

**Comment by atruskie** [comment]
Further comments:

-  Bad spacing on buttons
image
- modal dialog button should have class danger (be red)
- we can probably switch to an outline danger style for the abort buttons
    - reasoning: the abort buttons aren't immediately dangerous (they lead to a modal which has the actual danger button)

**Comment by hudson-newey** [comment]
I've revised the visual changes and updated the images of these changes in the PR description

File: src/app/components/harvest/pages/list/list.component.spec.ts [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+    // click the abort button
+    getAbortButton().click();
+    spec.detectChanges();
+
+    // click the cancel button
+    getModalNextButton().click();
+    spec.detectChanges();
+
+    // assert that transitionHarvest was *not* called
+    expect(harvestApi.transitionStatus).not.toHaveBeenCalledWith("complete");
```
Comment (atruskie) [inline src/app/components/harvest/pages/list/list.component.spec.ts]: this is incorrect??

we want the harvest to transition to complete if the abort button was clicked. 
```suggestion
    // assert that transitionHarvest *was* called
    expect(harvestApi.transitionStatus).toHaveBeenCalledWith("complete");
```

File: src/app/components/harvest/pages/list/list.component.spec.ts [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+    // ensure harvest status is not an abortable state
+    const unAbortableHarvest = new Harvest( generateHarvest({ status: "metadataExtraction" }) );
+
+    setup(
+      defaultProject,
+      unAbortableHarvest
+    );
+
+    // assert abort button is not rendered
+    expect(spec.query("#list-abort-button")).toBeFalsy();
```
Comment (atruskie) [inline src/app/components/harvest/pages/list/list.component.spec.ts]: should not be using a document id for a repeated element

File: src/app/components/harvest/pages/list/list.component.html [hunk 6]
```diff
@@ -42,6 +42,28 @@ <h1>Upload History</h1>
       <a class="btn btn-sm btn-primary" [bawUrl]="row.viewUrl">
         {{ asHarvest(row).status !== "complete" ? "Continue" : "View" }}
       </a>
+      <a
+        id="list-abort-button"
```
Comment (atruskie) [inline src/app/components/harvest/pages/list/list.component.html]: should not be using a an id for a repeated element

Comment (atruskie) [reply src/app/components/harvest/pages/list/list.component.html]: *maybe* try 
```suggestion
        name="abort-button"
```

instead

File: src/app/components/harvest/pages/list/list.component.ts [hunk 7]
```diff
@@ -44,11 +44,62 @@ class ListComponent extends PageComponent implements OnInit {
...
+      )
+      // We want this api request to complete regardless of lifecycle destruction
+      // eslint-disable-next-line rxjs-angular/prefer-takeuntil
+      .subscribe({
+        next: (): void => {
+          this.filters$.next({
+            sorting: {
+              direction: "desc",
+              orderBy: "createdAt"
+            }
```
Comment (atruskie) [inline src/app/components/harvest/pages/list/list.component.ts]: do we need to set this value? can we just call next with an empty object?

File: src/app/components/harvest/pages/list/list.component.ts [hunk 8]
```diff
@@ -44,11 +44,62 @@ class ListComponent extends PageComponent implements OnInit {
...
+  public isAbortable(harvest: Harvest): boolean {
+    const nonTransitionalStates: HarvestStatus[] = [
+      "complete",
+      "metadataExtraction",
+      "processing",
+      "scanning"
+    ];
+
+    return !nonTransitionalStates.includes(harvest.status);
+  }
```
Comment (atruskie) [inline src/app/components/harvest/pages/list/list.component.ts]: should be in the harvest model

File: src/app/components/harvest/screens/metadata-review/metadata-review.component.html [hunk 9]
```diff
@@ -75,33 +75,51 @@ <h3>Review</h3>
 </div>
 
 <div class="mt-3 d-flex justify-content-between">
-  <button
+  <div>
+    <button
+      id="harvest-abort-button"
```
Comment (atruskie) [inline src/app/components/harvest/screens/metadata-review/metadata-review.component.html:80]: is this id needed?

File: src/app/components/harvest/screens/metadata-review/metadata-review.component.html [hunk 10]
```diff
@@ -75,33 +75,51 @@ <h3>Review</h3>
...
+      class="btn btn-outline-danger me-2"
+      [disabled]="transitioningStage"
+      (click)="abortUpload(abortUploadModal)"
+    >
+      Abort
+    </button>
+
+    <button
+    id="harvest-upload-more-files-button"
```
Comment (atruskie) [inline src/app/components/harvest/screens/metadata-review/metadata-review.component.html]: is this id needed?

File: src/app/components/harvest/screens/metadata-review/metadata-review.component.html [hunk 11]
```diff
@@ -75,33 +75,51 @@ <h3>Review</h3>
...
-    [disabled]="transitioningStage"
-    (click)="processing(processingModal)"
-  >
-    {{
-      harvest.isMappingsDirty ? "Continue without checking changes" : "Continue"
-    }}
-  </button>
+  <div>
+    <button
+      id="harvest-check-changes-button"
```
Comment (atruskie) [inline src/app/components/harvest/screens/metadata-review/metadata-review.component.html]: why was this id added?

File: src/app/components/harvest/screens/metadata-review/metadata-review.component.spec.ts [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+    mocks: [ToastrService],
+  });
+
+  function setup() {
+    spec = createComponent({ detectChanges: false });
+
+    modalService = spec.inject(NgbModal);
+
+    stages = spec.inject<SpyObject<HarvestStagesService>>( HarvestStagesService as any );
+    stages.transition = jasmine.createSpy("transition") as any;
```
Comment (atruskie) [inline src/app/components/harvest/screens/metadata-review/metadata-review.component.spec.ts:61]: is there a reason this setup is different from the list spec?

File: src/app/components/harvest/screens/metadata-review/metadata-review.component.ts [hunk 13]
```diff
@@ -285,7 +294,7 @@ export class MetadataReviewComponent
     );
     this.tableLoading = false;
 
-    if (harvestItems.length === 0) {
+    if (harvestItems === undefined) {
```
Comment (atruskie) [inline src/app/components/harvest/screens/metadata-review/metadata-review.component.ts]: this seems like an incorrect change - the api should never be returning undefined for harvest items and now the condition no longer works for arrays of size 0. why was this changed?

File: src/app/directives/datatable/pagination.directive.ts [hunk 14]
```diff
@@ -185,7 +185,10 @@ export class DatatablePaginationDirective<Model extends AbstractModel>
     this.total$ = this.rows$.pipe(
       // Get the total number of models for the current filter from the
       // responses metadata
-      map((models): number => models[0]?.getMetadata().paging.total ?? 0)
+      map((models): number => {
+        console.log(models[0], models[0].getMetadata());
```
Comment (atruskie) [inline src/app/directives/datatable/pagination.directive.ts]: this log line should not have been committed, revert this change

File: src/app/services/baw-api/harvest/harvest.service.ts [hunk 15]
```diff
@@ -104,6 +104,17 @@ export class HarvestsService implements StandardApi<Harvest, [IdOr<Project>]> {
...
+  public isAbortable(
+    model: Harvest
+  ): boolean {
+    const transitionableStates: HarvestStatus[] = [
+      "metadataReview",
+      "uploading"
+    ];
+
+    return transitionableStates.includes(model.status);
+  }
```
Comment (atruskie) [inline src/app/services/baw-api/harvest/harvest.service.ts]: this should be in the harvest model, not the service

File: src/app/services/baw-api/harvest/harvest.service.spec.ts [hunk 16]
```diff
@@ -33,4 +37,26 @@ describe("HarvestsService", () => {
...
+        const mockHarvest = new Harvest(
+          generateHarvest({
+            status: harvestStatus
+          })
+        );
+
+        expect(spec.service.isAbortable(mockHarvest)).toBeTrue();
+      });
+    });
+  });
```
Comment (atruskie) [inline src/app/services/baw-api/harvest/harvest.service.spec.ts]: this should be in the model specs not in the service specs

File: src/app/services/baw-api/harvest/harvest.service.spec.ts [hunk 17]
```diff
@@ -33,4 +37,26 @@ describe("HarvestsService", () => {
     harvestId, // harvest
     5 //project
   );
+
+  describe("isAbortable", () => {
+    const harvestTransitionalStates: HarvestStatus[] = [
+      "metadataReview",
+      "uploading"
+    ];
```
Comment (atruskie) [inline src/app/services/baw-api/harvest/harvest.service.spec.ts]: you don't test the other stages?

try making two sets - one with allowed states and one with disallowed states

File: src/app/services/baw-api/harvest/harvest.service.spec.ts [hunk 18]
```diff
@@ -33,4 +37,26 @@ describe("HarvestsService", () => {
...
+
+    harvestTransitionalStates.forEach((harvestStatus: HarvestStatus) => {
+      it(`should return true for a Harvest with the status of ${harvestStatus}`, () => {
+        defaultProject = new Project(generateProject());
+        defaultProject.addMetadata(generateProjectMeta({}));
```
Comment (atruskie) [inline src/app/services/baw-api/harvest/harvest.service.spec.ts]: these seem to be entirely redundant?

File: src/app/test/fakes/Harvest.ts [hunk 19]
```diff
@@ -53,9 +53,7 @@ export function generateHarvest(data?: Partial<IHarvest>): Required<IHarvest> {
     uploadPassword: modelData.internet.password(),
     uploadUser: modelData.internet.userName(),
     uploadUrl: `sftp://${modelData.random.word()}.${modelData.random.word()}.org:2020`,
-    mappings: Array(modelData.datatype.number({ min: 0, max: 3 })).map(() =>
-      generateHarvestMapping()
-    ),
+    mappings: Array(generateHarvestMapping()),
```
Comment (atruskie) [inline src/app/test/fakes/Harvest.ts]: is there a reason for this change? It seems unrelated to the work you were doing

**Review by atruskie — CHANGES_REQUESTED** [review]

File: src/app/components/harvest/screens/metadata-review/metadata-review.component.ts (hunk 13 above)
Comment (hudson-newey) [reply src/app/components/harvest/screens/metadata-review/metadata-review.component.ts]: I've reverted this change. It's an artifact from my testing

File: src/app/components/harvest/pages/list/list.component.spec.ts (hunk 4 above)
Comment (hudson-newey) [reply src/app/components/harvest/pages/list/list.component.spec.ts]: I have fixed this

File: src/app/components/harvest/screens/metadata-review/metadata-review.component.html (hunk 9 above)
Comment (hudson-newey) [reply src/app/components/harvest/screens/metadata-review/metadata-review.component.html:80]: It's used in the unit tests, but not in page functionality

File: src/app/components/harvest/screens/metadata-review/metadata-review.component.html (hunk 10 above)
Comment (hudson-newey) [reply src/app/components/harvest/screens/metadata-review/metadata-review.component.html]: This id is not needed, I have since removed it

File: src/app/components/harvest/screens/metadata-review/metadata-review.component.html (hunk 9 above)
Comment (atruskie) [reply src/app/components/harvest/screens/metadata-review/metadata-review.component.html:80]: that is an excellent reason for it not to be there

ids are gloabal identifiers attached to a document

they should not be repeated and should not be used unless needed

File: src/app/components/harvest/screens/metadata-review/metadata-review.component.html (hunk 11 above)
Comment (hudson-newey) [reply src/app/components/harvest/screens/metadata-review/metadata-review.component.html]: This id is not necessary. I have since reverted this change

File: src/app/components/harvest/pages/list/list.component.ts (hunk 7 above)
Comment (hudson-newey) [reply src/app/components/harvest/pages/list/list.component.ts]: Calling with an empty object does force the filters$ to update
I have since made the necessary code changes

File: src/app/services/baw-api/harvest/harvest.service.spec.ts (hunk 18 above)
Comment (hudson-newey) [reply src/app/services/baw-api/harvest/harvest.service.spec.ts]: I 100% agree

I have since removed it

File: src/app/services/baw-api/harvest/harvest.service.ts (hunk 15 above)
Comment (hudson-newey) [reply src/app/services/baw-api/harvest/harvest.service.ts]: Done

File: src/app/test/fakes/Harvest.ts (hunk 19 above)
Comment (hudson-newey) [reply src/app/test/fakes/Harvest.ts]: The mapping attribute was causing troubles for me when I was developing my tests
This change is no longer needed since the tests are now complete

I have since removed the testing artefact

File: src/app/components/harvest/screens/metadata-review/metadata-review.component.spec.ts (hunk 12 above)
Comment (hudson-newey) [reply src/app/components/harvest/screens/metadata-review/metadata-review.component.spec.ts:61]: This spec was created before the list spec

The spy's on the list spec are not working properly, so I'm considering lifting this setup from metadata-review to the list spec since I know that the setup on metadata-review as expected. This would also work well for consistency & continuity, as the batch-upload component uses a similar transitional spy setup as the metadata-review component spec

File: src/app/models/Harvest.ts [hunk 20]
```diff
@@ -167,6 +167,14 @@ export class Harvest extends AbstractModel implements IHarvest {
     );
   }
 
+  public isAbortable(): boolean {
+    const transitionableStates: HarvestStatus[] = [
+      "metadataReview",
+      "uploading"
```
Comment (atruskie) [inline src/app/models/Harvest.ts]: this incorrect, there was a reason i gave you the link to the server code

**Review by atruskie — CHANGES_REQUESTED** [review]
the ubernested callbacks an interval waiting periods are dodgy - try waiting for the actual events

**Review by atruskie — APPROVED** [review]
Glad we got this done. Nice work. 

Flatten the commits and then merge when you're ready.
